From ba01e2052f25ce40537dcac35769d5d70ef5e196 Mon Sep 17 00:00:00 2001
From: Black-Seraph <admin@black-seraph.com>
Date: Thu, 11 May 2023 15:18:07 +0200
Subject: [PATCH] Add period size override support

---
 audio/drivers/tinyalsa.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/audio/drivers/tinyalsa.c b/audio/drivers/tinyalsa.c
index b16f176..14f09e5 100644
--- a/audio/drivers/tinyalsa.c
+++ b/audio/drivers/tinyalsa.c
@@ -96,10 +96,17 @@ static void * tinyalsa_init(const char *devicestr, unsigned rate, unsigned laten
 	if (orig_rate != rate)
 		*new_rate = rate;
 
+	unsigned int period_size = 1024;
+	char * period_size_override = getenv("ALSA_DEFAULT_PERIOD_SIZE");
+	if (period_size_override != NULL)
+	{
+		period_size = (unsigned int)atoi(period_size_override);
+	}
+
 	config.rate = rate;
 	config.format = is_little_endian() ? PCM_FORMAT_S16_LE : PCM_FORMAT_S16_BE;
 	config.channels = 2;
-	config.period_size = 1024;
+	config.period_size = period_size;
 	config.period_count = 4;
 	config.start_threshold = config.period_size;
 	config.stop_threshold = config.period_size * config.period_count;
-- 
2.25.1

