From e48d4d1bfb7d655269065f30729b111443750f12 Mon Sep 17 00:00:00 2001
From: Black-Seraph <admin@black-seraph.com>
Date: Fri, 12 May 2023 18:15:10 +0200
Subject: [PATCH] Allow forcing of MMAP in tinyalsa

---
 audio/drivers/tinyalsa.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/audio/drivers/tinyalsa.c b/audio/drivers/tinyalsa.c
index 108eaa6..63c2aa7 100644
--- a/audio/drivers/tinyalsa.c
+++ b/audio/drivers/tinyalsa.c
@@ -50,6 +50,7 @@ typedef struct tinyalsa
 	struct pcm *pcm;
 	size_t buffer_size;
 	bool nonblock;
+	bool use_mmap;
 } tinyalsa_t;
 
 static void * tinyalsa_init(const char *devicestr, unsigned rate, unsigned latency, unsigned block_frames, unsigned *new_rate)
@@ -72,6 +73,7 @@ static void * tinyalsa_init(const char *devicestr, unsigned rate, unsigned laten
 
 	tinyalsa->card = card;
 	tinyalsa->device = device;
+	tinyalsa->use_mmap = getenv("ALSA_FORCE_MMAP") != NULL;
 
 	RARCH_LOG("[TINYALSA]: Using card: %u, device: %u.\n", card, device);
 
@@ -116,7 +118,7 @@ static void * tinyalsa_init(const char *devicestr, unsigned rate, unsigned laten
 	tinyalsa->config.silence_threshold = tinyalsa->config.period_size * tinyalsa->config.period_count;
 	tinyalsa->config.silence_size = 0;
 
-	tinyalsa->pcm = pcm_open(card, device, PCM_OUT | PCM_MMAP | PCM_NONBLOCK, &tinyalsa->config);
+	tinyalsa->pcm = pcm_open(card, device, PCM_OUT | PCM_NONBLOCK | (tinyalsa->use_mmap ? PCM_MMAP : 0), &tinyalsa->config);
 
 	if (!tinyalsa->pcm)
 	{
@@ -240,7 +242,7 @@ static bool tinyalsa_start(void *data, bool is_shutdown)
 
 	if (tinyalsa->pcm == NULL)
 	{
-		tinyalsa->pcm = pcm_open(tinyalsa->card, tinyalsa->device, PCM_OUT | PCM_MMAP | PCM_NONBLOCK, &tinyalsa->config);
+		tinyalsa->pcm = pcm_open(tinyalsa->card, tinyalsa->device, PCM_OUT | PCM_NONBLOCK | (tinyalsa->use_mmap ? PCM_MMAP : 0), &tinyalsa->config);
 	}
 
 	return tinyalsa->pcm != NULL;
-- 
2.25.1

