From 4bfcf9ce7664e9a3c2169f705d3b03d2cd14b0ef Mon Sep 17 00:00:00 2001
From: turtleletortue <46635000+turtleletortue@users.noreply.github.com>
Date: Sun, 26 Nov 2023 09:46:46 -0500
Subject: [PATCH 10/10] Add video and input backend override quirks

---
 retroarch.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/retroarch.c b/retroarch.c
index f4c7a6a1e9..75dc33f615 100644
--- a/retroarch.c
+++ b/retroarch.c
@@ -7691,6 +7691,24 @@ bool retroarch_main_init(int argc, char *argv[])
       }
    }
 
+   // Get the default backend quirks
+   char * video_backend = getenv("RETROARCH_DEFAULT_VIDEO_BACKEND");
+   char * input_backend = getenv("RETROARCH_DEFAULT_INPUT_BACKEND");
+
+   // We've been asked to override the video backend
+   if (video_backend != NULL)
+   {
+      // Override video driver
+      configuration_set_string(settings, settings->arrays.video_driver, video_backend);
+   }
+
+   // We've been asked to override the input backend
+   if (input_backend != NULL)
+   {
+      // Override input driver
+      configuration_set_string(settings, settings->arrays.input_driver, input_backend);
+   }
+
    /* Pre-initialize all drivers
     * Attempts to find a default driver for
     * all driver types.
-- 
2.34.1

